lib/3scale/backend/job_fetcher.rb

```diff
@@ -11,49 +11,41 @@ class JobFetcher
         @redis = redis_client
         @fetch_timeout = fetch_timeout
         @queues ||= QUEUES.map { |q| "queue:#{q}" }
       end
 
-      def fetch
-        encoded_job = pop_from_queue
-        return nil if encoded_job.nil? || encoded_job.empty?
-
-        begin
-          # Resque::Job.new accepts a queue name as a param. It is very
-          # important to set here the same name as the one we set when calling
-          # Resque.enqueue. Resque.enqueue uses the @queue ivar in
-          # BackgroundJob classes as the name of the queue, and then, it stores
-          # the job in a queue called resque:queue:_@queue_. 'resque:' is the
-          # namespace and 'queue:' is added automatically. That's why we need
-          # to call blpop on 'queue:#{q}' above. However, when creating the job
-          # we cannot set 'queue:#{q}' as the name. Otherwise, if it fails and
-          # it is re-queued, it will end up in resque:queue:queue:_@queue_
-          # instead of resque:queue:_@queue_.
-          encoded_job[0].sub!('queue:', '')
-          Resque::Job.new(encoded_job[0],
-                          Yajl::Parser.parse(encoded_job[1], check_utf8: false))
-        rescue Exception => e
-          # I think that the only exception that can be raised here is
-          # Yajl::ParseError. However, this is a critical part of the code so
-          # we will capture all of them just to be safe.
-          Worker.logger.notify(e)
-          nil
-        end
+      # @param wait [Boolean] wait for a job to appear in a queue instead of returning immediately
+      # @param max [Integer] maximum number of records to return
+      #   note: presently ignored in waiting mode because BLMPOP was added in Redis 7
+      #   note: also we need to update redis-rb gem so that sync client can support this parameter
+      # @return [nil, Resque::Job, Array<Resque::Job>] nil when no jobs are present in the queue,
+      #                                                 a single job when max was not specified,
+      #                                                 an array of jobs otherwise
+      def fetch(wait: true, max: nil)
+        encoded_jobs = wait ? wait_pop_from_queue : try_pop_from_queue(max)
+        return if !encoded_jobs || encoded_jobs.empty?
+
+        jobs = encoded_jobs.map { decode_job _1 }
+
+        jobs.count > 1 || max ? jobs : jobs.first
       end
 
       # Note: this method calls #close on job_queue after receiving #shutdown.
       # That signals to the caller that there won't be any more jobs.
       def start(job_queue)
         loop do
           break if @shutdown
-          job = fetch
-          job_queue << job if job
+
+          jobs = fetch(wait: false, max: Worker::DEFAULT_MAX_CONCURRENT_JOBS)
+
+          # if there were no jobs, we can make a blocking call instead of reinventing waiting logic
+          jobs = fetch(max: Worker::DEFAULT_MAX_CONCURRENT_JOBS) unless jobs
```

Comment:
I don't see the point on `wait:false` and this double call to `fetch`. First we call it non-waiting, if it returns nothing because there's nothing in any queue, then wait. OK, But that's what will happen anyway if we call it with `wait:true` directly, isn't it? So the first call seems redundant to me, and the `wait` parameter unnecessary.

Reply:
See my other comment. `lpop` fetch multiple values with one call so we can fetch much more efficient.